### PR TITLE
fix(levm): avoid jumpdest in push args

### PR DESCRIPTION
**Pull request by maximopalopoli**
**Motivation**

The motivation is to do not count the push arguments in the valid jump destinations.

**Description**

This is done by calculating the valid jump destinations when the vm is initialized. This PR also adds the valid_jump_destinations attribute to callframe, and uses it to determine valid jumps in is_valid_jump_addr().
This PR has its origin in making the vmIOandFlowOperations tests pass, and also is inspired in [execution-specs implementation](https://github.com/ethereum/execution-specs/) about valid jump destinations.

Note: These changes caused me problems with the benchmarks, since they created the vm with an empty bytecode and then set a new callframe outside of the new one. What I ended up doing is that the init of the callframe in which a bytecode is passed also calculates the valid jump dests, which makes the benchmarks go a bit slower.

### CODE REVIEW

**Review by ilitteri — APPROVED** [review]
LGTM

**Review by JereSalo — CHANGES_REQUESTED** [review]
I made some suggestions. Some of them are more optional than others. It's up to you to choose.

File: crates/vm/levm/src/call_frame.rs [hunk 1]
```diff
@@ -103,6 +107,7 @@ impl CallFrame {
         gas_limit: U256,
         gas_used: U256,
         depth: usize,
+        valid_jump_destinations: HashSet<usize>,
```
[the remark this answers is not in the document]
Comment (maximopalopoli) [reply crates/vm/levm/src/call_frame.rs]: Done in [Move valid jump destinations to Callframe::new](https://github.com/lambdaclass/ethrex/pull/1434/commits/a6d9e2edcf2fd1e64264c1338603b135b6022809)!

File: crates/vm/levm/src/vm.rs [hunk 2]
```diff
@@ -71,6 +71,43 @@ pub fn word_to_address(word: U256) -> Address {
...
+    let mut valid_jump_destinations = HashSet::new();
+    let mut pc = 0;
+
+    while pc < code.len() {
+        let code_position = match code.get(pc) {
+            Some(code) => *code,
+            None => {
+                break;
+            }
+        };
```
[the remark this answers is not in the document]
Comment (maximopalopoli) [reply crates/vm/levm/src/vm.rs]: Done in [change while+match for while let some](https://github.com/lambdaclass/ethrex/pull/1434/commits/6389b71cb0d4de4aa35975ef95c68bc624f2e19c)

File: crates/vm/levm/src/call_frame.rs [hunk 3]
```diff
@@ -80,15 +81,23 @@ pub struct CallFrame {
...
-            gas_limit: U256::MAX,
-            ..Default::default()
-        }
+            Default::default(),
+            Default::default(),
+            Default::default(),
+            U256::MAX,
+            Default::default(),
+            Default::default(),
+        )
```
Comment (ilitteri) [inline crates/vm/levm/src/call_frame.rs:100]: You don't need to write all those `Default::default()`s, you can keep the previous implementation and just add the bytecode field.

Comment (maximopalopoli) [reply crates/vm/levm/src/call_frame.rs]: Done in [Return to new_from_bytecode previous version](https://github.com/lambdaclass/ethrex/pull/1434/commits/00f194205f4db0d4dfa8e155e355f68be71ecfa1)

File: crates/vm/levm/src/vm.rs [hunk 4]
```diff
@@ -71,6 +71,43 @@ pub fn word_to_address(word: U256) -> Address {
...
+        let current_opcode = Opcode::from(code_position);
+
+        if current_opcode == Opcode::JUMPDEST {
+            // If current opcode is jumpdest, add it to valid destinations set
+            valid_jump_destinations.insert(pc);
+        } else if (Opcode::PUSH1..=Opcode::PUSH32).contains(&current_opcode) {
+            // If current opcode is push, skip as many positions as the size of the push
+            let pushed_size = code_position
+                .checked_sub(u8::from(Opcode::PUSH1))
+                .ok_or(VMError::InvalidBytecode)?;
```
[the remark this answers is not in the document]
Comment (maximopalopoli) [reply crates/vm/levm/src/vm.rs]: Done in [change for internal + rename var](https://github.com/lambdaclass/ethrex/pull/1434/commits/b479b919becca5978e4ef175aefa4df9c43f47f2)

File: crates/vm/levm/src/vm.rs [hunk 5]
```diff
@@ -71,6 +71,43 @@ pub fn word_to_address(word: U256) -> Address {
...
+        };
+
+        let current_opcode = Opcode::from(code_position);
+
+        if current_opcode == Opcode::JUMPDEST {
+            // If current opcode is jumpdest, add it to valid destinations set
+            valid_jump_destinations.insert(pc);
+        } else if (Opcode::PUSH1..=Opcode::PUSH32).contains(&current_opcode) {
+            // If current opcode is push, skip as many positions as the size of the push
+            let pushed_size = code_position
```
[the remark this answers is not in the document]
Comment (maximopalopoli) [reply crates/vm/levm/src/vm.rs]: Done in [change for internal + rename var](https://github.com/lambdaclass/ethrex/pull/1434/commits/b479b919becca5978e4ef175aefa4df9c43f47f2)

File: crates/vm/levm/src/vm.rs [hunk 6]
```diff
@@ -71,6 +71,43 @@ pub fn word_to_address(word: U256) -> Address {
...
+                .checked_sub(u8::from(Opcode::PUSH1))
+                .ok_or(VMError::InvalidBytecode)?;
+            let skip_length =
+                usize::from(pushed_size.checked_add(1).ok_or(VMError::InvalidBytecode)?);
```
[the remark this answers is not in the document]
Comment (maximopalopoli) [reply crates/vm/levm/src/vm.rs]: [Also change to Internal overflow](https://github.com/lambdaclass/ethrex/pull/1434/commits/b1148986178d0e81a43774e835ab585e897673bf)

**Review by JereSalo — APPROVED** [review]
Good job!

**Pull request merged**